Hybrid recurrent cache

gabe-l-hart opened the pull request:
## Dependencies

- [ ] https://github.com/ggml-org/llama.cpp/pull/13746

## Description

This PR replaces my initial (more generic) attempt in https://github.com/ggml-org/llama.cpp/pull/13276. Following [Georgi's comment](https://github.com/ggml-org/llama.cpp/pull/13746/files#r2110971927), this version is explicit about owning two child caches, one `unified` and one `recurrent`. This greatly simplifies the logic and avoids the need for additional interface changes to support managing child caches in the abstract.

This PR has been verified to be minimally functional in #13550 (Granite 4 / Bamba).

gabe-l-hart committed:
feat: Add support for distinguishing recurrent vs non-recurrent layers in hparams

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add c++ side constants for attention layer indices hparam

Branch: GraniteFour

gabe-l-hart committed:
feat: Add llama_model_is_hybrid API call

Also, split llama_model_is_recurrent into llm_arch_is_recurrent in
llama-arch with llama_model_is_recurrent delegating to
llm_arch_is_recurrent. The same split is done for hybird. This is needed
because there are places where the llama_model has not yet been initialized
but we need to check if the model is recurrent (specifically for the
per-layer recurrent check array in hparams).

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Auto-fill hparams.recurrent_layer_arr based on whether the model is recurrent

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Use per-layer sizing everywhere in kv caches

Branch: GraniteFour

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Move layer_filter_cb up to llama_kv_cache

This will be needed by other cache types as well, so centralizing the
definition will make it more reusable.

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Add layer filter to recurrent cache

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
fix: Fix indexing into k_l for recurrent cache with filter

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ggerganov committed:
kv-cache : simplify the "struct llama_kv_cache" interface

ggml-ci

ggerganov committed:
kv-cache : revert the (n_swa + n_ubatch) change (for next PR)

ggml-ci

ggerganov committed:
kv-cache : some comments

ggml-ci

ggerganov committed:
context : fix graph reserve for multiple sequences

ggml-ci

ggerganov committed:
kv-cache : fix typo [no ci]

ggerganov committed:
kv-cache : fix find_slot() logic for free slots

ggml-ci

ggerganov committed:
llama : add TODO for deprecating the defrag API in the future

ggerganov committed:
kv-cache : improve find_slot() using min/max seq pos info

ggml-ci

ggerganov committed:
llama : handle aborts and compute errors

ggml-ci

gabe-l-hart committed:
refactor: rename *_is_hybrid -> *_is_hybrid_recurrent

The implementation of the hybrid cache intentionally does not specify the
types of the child caches, so there was a naming mismatch with these
predicate functions that used "hybrid" to imply "hybrid recurrent."

Branch: HybridCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Construct hybrid recurrent cache for hybrid recurrent models

This includes a refactor of the create_memory logic to avoid needing to use
the arch enum explicitly unless a model needs explicit cache instantiation
logic beyond the standard logic for recurrent, hybrid, unified, and iswa.

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ggerganov committed:
memory : extract state into llama_memory_state

ggml-ci

gabe-l-hart committed:
fix: Fix wrong bool condition for split equal in hybrid cache

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

ggerganov committed:
kv-cache : add comments

ggml-ci

ggerganov commented:
Sorry about the massive changes, hopefully we are almost converging to the final version of the implementation. See https://github.com/ggml-org/llama.cpp/pull/13746#issuecomment-2921703899 for the latest short-term plan.

gabe-l-hart commented:
No problem, and thanks for the quick work getting these pieces overhauled! I'll plan to keep this branch up to date as the changes roll in so that it will hopefully be ready for review quickly once its turn comes.

gabe-l-hart committed:
feat: First pass at llama_kv_cache_hybrid_recurrent

This follows the pattern in iswa where the two child caches are held
explicitly to support the case where a model requires a single attention
cache and a single recurrent cache where each layer uses exactly one of the
caches.

This is a rewrite of the more generic approach in the original hybrid cache
PR: https://github.com/ggml-org/llama.cpp/pull/13276

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart committed:
feat: Support hybrid recurrent cache in llm_graph_context

Branch: HybridRecurrentCache

Signed-off-by: Gabe Goodhart <ghart@us.ibm.com>

gabe-l-hart commented:
Since this was auto-closed when #13746 merged, I've rebased and reopened as https://github.com/ggml-org/llama.cpp/pull/13979
